File: 543-diameter-of-binary-tree/diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int find(TreeNode* root){
        if(root==NULL) return 0;
        return 1+max(find(root->left),find(root->right));
    }
    int diameterOfBinaryTree(TreeNode* root) {
        if(root==NULL) return 0;
        int a=diameterOfBinaryTree(root->left);
        int b=diameterOfBinaryTree(root->right);
        int right=find(root->left);
        int left=find(root->right);
        return max({a,b,left+right});        
    }
};

File: 543-diameter-of-binary-tree/diameter-of-binary-tree.java
class Solution {
    int res;
    public int find(TreeNode root){
        if(root==null) return 0;

        int left=find(root.left);
        int right=find(root.right);

        if(res<left+right){
            res=left+right;
        }
        return 1+Math.max(left,right);
    }
    public int diameterOfBinaryTree(TreeNode root) {
        res=0;
        find(root);
        return res;
    }
}